scipy/ndimage/src/nd_image.c

```diff
@@ -934,6 +934,227 @@ static PyObject *Py_FindObjects(PyObject *obj, PyObject *args)
     }
 }
 
+/*
+   Implement the ndimage.value_indices() function.
+   Makes 3 passes through the array data. We use ndimage's NI_Iterator to
+   iterate through all elements of the input array.
+
+   To support all numpy datatypes, we define several macros which take a
+   parameter for the array datatype in question. By using these, all comparisons
+   are carried out using the same datatype as the input array.
+*/
+#define VALUEINDICES_MINVAL(valType) (*((valType *)PyArray_GETPTR1(minMaxArr, 0)))
+#define VALUEINDICES_MAXVAL(valType) (*((valType *)PyArray_GETPTR1(minMaxArr, 1)))
+#define VALUEINDICES_IGNOREVAL(valType) (*((valType *)PyArray_GETPTR1(ignorevalArr, 0)))
+#define CASE_VALUEINDICES_SET_MINMAX(valType) {\
+    valType val = *((valType*)arrData); \
+    if (nullIsNone || (val != VALUEINDICES_IGNOREVAL(valType))) {\
+        if (minMaxUnset) { \
+            VALUEINDICES_MINVAL(valType) = val;  \
+            VALUEINDICES_MAXVAL(valType) = val; \
+            minMaxUnset = 0; \
+        } else { \
+            if (val < VALUEINDICES_MINVAL(valType)) VALUEINDICES_MINVAL(valType) = val; \
+            if (val > VALUEINDICES_MAXVAL(valType)) VALUEINDICES_MAXVAL(valType) = val; \
+        } \
+    }\
+}
+#define CASE_VALUEINDICES_MAKEHISTOGRAM(valType) {\
+    numPossibleVals = (VALUEINDICES_MAXVAL(valType) - VALUEINDICES_MINVAL(valType) + 1); \
```

Comment:
Are we worried about memory allocation for uint64 and similar arrays? Do we need to compute unique values first?

```
vals = np.unique(x)
hist_indices = {v: i for (v, i) in zip(vals, range(len(vals))}
```

Reply:
Thanks @stefanv 
I did briefly consider this matter when writing this code, but you have prompted me to do a detailed analysis. It did not change my conclusion, but good to capture the reasoning that supports to this design. 

**Setting the scene**
    • For cases where the range of values is relatively small, it doesn’t matter much, obviously.
    • For cases where the range between minval and maxval is very large (into millions), the distinction you are raising is whether the range is densely or sparsely used. 
    • I believe (without hard data, just experience) that the far more common case would be for a large range to be densely used. I can’t think of situations where a small number of values would be spread over such a large range. It might arise, but it seems to me it would be much less common. The situations which motivated this work would all have contiguous values. 
    • Just for completeness - as I am sure you know, in addition to the hist array, exactly the same issues apply to the valCtr and ndxPtr arrays. 

**Detailed reasoning**
In the current implementation, the memory usage of both the sparse and dense cases is identical. If we changed to use dictionaries for these internal arrays, the memory usage would indeed be much better for the sparse case, but would be much worse for the dense case. Some tests using sys.getsizeof() suggest that a densely populated dictionary with keys and values being int64 occupies 4 to 5 times the memory as an equivalent array. 

Assuming I am correct about which is the more common case, this would mean that the memory usage of the common case is made worse in order to improve usage of the rare case, whereas the current implementation would perform equally well in rare and common cases.

There are also some speed implications, as the dictionary lookups ought to be slower. However, I am not sure whether this would be serious, as Python dictionary lookups with integer keys are quite cleverly handled. It would probably be quite dependent on the range in question, but in principle it would degrade with a larger range of values (I think… ). 

It is perhaps also worth noting that these kind of spare/dense trade-offs do cause failures elsewhere. If one tried to use numpy.histogram() to do this step, it would suffer from the same problem. Somewhat similarly, as discussed above, the skimage regionprops code suffers from memory blow-out if the different value regions are not spatially compact, and again, this is probably not a serious issue in normal usage. 

So, after walking through all those aspects, I have concluded that the current implementation is probably the best one, on balance. Does all that make sense to you?

Reply:
> * I can’t think of situations where a small number of values would be spread over such a large range.

The typical situation is where a specific value is used as a NULL value, typically max(range).

For now, why don't we get this version in and see if anyone requests the sparse feature?